lib/pbench/test/functional/agent/cli/commands/test_config.py

```diff
@@ -0,0 +1,40 @@
+import os
+import pathlib
+import pytest
+
+from pbench.agent.base import BaseCommand
+
+
+def test_pbench_run(monkeypatch, tmpdir, agent_config, pbench_run):  # noqa F811
+    """Use pbench_run value from pbench-agent.cfg file"""
+    monkeypatch.delenv("pbench_run")
+    env_pbench_run = os.environ.get("pbench_run")
+    assert env_pbench_run is None
+    command = ["pbench-list-tools"]
+    out, err, exitcode = pytest.helpers.capture(command)
+    assert exitcode == 0
+    assert pathlib.Path(f"{tmpdir}/var/lib/pbench-agent") == pbench_run
+
+
+def test_pbench_run_dir_existence(monkeypatch, agent_config, pbench_run):
+    """Verify the existence of a non-existing pbench_run directory
+    which results in expected failure of the test"""
+    monkeypatch.setenv("pbench_run", f"{pbench_run}/test")
+    command = ["pbench-list-tools"]
+    out, err, exitcode = pytest.helpers.capture(command)
+    assert exitcode == 1
+    assert b"" == out
+    assert b"[ERROR] pbench run directory does not exist," in err
+
+
+def test_pbench_dir_existence(monkeypatch, pbench_run, agent_config):
+    """Verify the existence of existing pbench_run directory
+    and successful completion of command"""
+    pbench_run_d = pbench_run / "test_dir"
+    pbench_run_d.mkdir(parents=True, exist_ok=True)
+    monkeypatch.setenv("pbench_run", str(pbench_run_d))
+    command = ["pbench-list-tools"]
+    out, err, exitcode = pytest.helpers.capture(command)
+    pbench_run_val = BaseCommand.pbench_run
```

Comment:
BaseCommand doesn't have a class attribute `pbench_run`, its an object property. You may have to create BaseCommand object to refer to the pbench_run attribute

Reply:
That is an abstract class due to which direct call or object creation is not possible. Calling it indirectly also require `context` will look for some other way

Reply:
Yeah, this is a consequence of the way we run our `click` commands in test: the `capture` helper fires up the whole command in a `Popen` subprocess and captures the results... so we have no access to any of the internal command context to check. We can only rely on `stderr`, `stdout`, `proc.returncode`, and the state of the `tmp` filesystem tree. 

So in this test we can only verify that we saw the tools we expected and whether the command succeeded or failed. We have to pretty much trust that it connected inputs and outputs the way we hoped.

That's what makes this a "functional" test, but like the legacy gold tests it's a hacked environment so it's neither a unit test nor a true functional test. I'd love to get rid of this whole category: both for agent and server. ☹️

Reply:
As Dave points out, the test style in use here is a "functional test" model, where we actually invoke a real tool and try to make sure that it functions correctly under various circumstances.

The "unit test" model would be to invoke only the code which we want to test, here that would be `BaseCommand.__init__()`.  We _could_ do that directly, but that's a bit complicated, since it's an object constructor.  Typically, we would test those by instantiating the object, but this is an abstract base class, so we cannot create an instance of it.  So, to address that, we would have the unit test code declare a concrete subclass of it and instantiate that.

The challenge is that the unit test will need to come up with a Click context object to pass to the instantiation...but maybe that's not too hard.  This being Python, I think that all we need is a generic object which has a field named `config` with an appropriate value in it.  So, maybe something like this would work:
```
    class TestCommand(BaseCommand):
        class FakeContext:
            config = "/path/to/config_file"

        def __init__(self):
            super(TestCommand, self).__init__(FakeContext())

    pbench_run_val = TestCommand().pbench_run
```

If that works well, you might consider using it in the other tests to make them more unit-ish and less functinal-ish.  (To make them properly unit-ish, you'd probably want to mock `PbenchAgentConfig.__init__()` and maybe a few other things, but this would be a good start.)

Reply:
I think I'd rather not put "starting to move towards unit testing" under *this* PR, and simply conform to the current pseudo-functional testing style for now. (For one thing, let's not contemplate moving these from `test/functional` to `test/unit`.)

That means we can't measure internal aspects of the environment that aren't externally observable. That's OK; at this level, if there's no externally observable effect then it simply doesn't matter.

(However, it ought to be observable: if we re-direct `pbench_run` we shouldn't find the tools registered under the (temp) `/var/lib/pbench_agent` and should instead find tools registered under our `pbench_run` directory. Right?)

Reply:
By the way, in case it's not obvious...if you switch to a unit testing model (where you get the value of `pbench_run` directly from your own instantiation of a subclass of `BaseCommand`), then you don't need to invoke `capture()` or to run `pbench-list-tools` at all, because you'll be exercising the CUT directly.

Reply:
> I think I'd rather not put "starting to move towards unit testing" under this PR, and simply conform to the current pseudo-functional testing style for now. (For one thing, let's not contemplate moving these from test/functional to test/unit.)

Well, given that these tests are brand new code, they don't have to be put "here", they could be put wherever its appropriate (from the perspective of the `main` branch, that's a "different addition", not a "move").

Nevertheless, I don't disagree that it would be best to have them conform to the approach in the surrounding code/modules...so, if they are going to stay in the functional area, it would probably be best if they stayed "functional".  And, toward that end, I concur that their results should be observable.  So, rather than running (just) `pbench-list-tools` we should have the tests run `pbench-register-tools` and then check to see that the tools end up in the proper location.

But, I still think it would be better to convert these to a more unit-ish model and then put them in the appropriate place (e.g., add them to the existing `lib/pbench/test/unit/agent/test_config.py`).

Reply:
Maybe what I can do is, Open another issue, (I think there are issues available for this) to move away from Functional tests and leave this as it is. In the third test, I can register the tools in the given location with `pbench-register-tool` and check whether the tools are registered in the given location with the help of `pbench-list-tools`.

Reply:
Hey @riya-17, I'll mark this as resolved, but can you please mention one of those issues here for our tracking.